llpc/lower/llpcSpirvLowerMath.cpp

```diff
@@ -295,16 +283,154 @@ bool SpirvLowerMathPrecision::runImpl(Module &module) {
         valueWritten = callInst->getOperand(0);
       }
 
-      if (valueWritten && builtIn == lgc::BuiltInPosition && enableImplicitInvariantExports) {
+      if (valueWritten && builtIn == lgc::BuiltInPosition) {
         disableFastMath(valueWritten);
         changed = true;
       }
     }
   }
+  return changed;
+}
+
+static bool clearContractFlag(Instruction *inst) {
+  if (!isa<FPMathOperator>(inst))
+    return false;
+  LLVM_DEBUG(dbgs() << "clearing contract flags: " << *inst << "\n");
+  FastMathFlags fastMathFlags = inst->getFastMathFlags();
+  fastMathFlags.setAllowReassoc(false);
+  fastMathFlags.setAllowContract(false);
+  inst->copyFastMathFlags(fastMathFlags);
+  return true;
+}
+
+bool SpirvLowerMathPrecision::propagateNoContract(Module &module, bool forward, bool backward) {
+  bool changed = false;
+
+  SmallVector<Instruction *> roots;
+  DenseSet<Instruction *> visited;
+
+  // Find all NoContract instructions to build root set
+  LLVM_DEBUG(dbgs() << "locate no contract roots\n");
+  for (auto &func : module) {
+    for (auto &block : func) {
+      for (auto &inst : block) {
+        if (isNoContract(&inst)) {
+          LLVM_DEBUG(dbgs() << "root: " << inst << "\n");
+          roots.push_back(&inst);
+          visited.insert(&inst);
+        }
+      }
+    }
+  }
+
+  SmallVector<Instruction *> worklist;
+
+  // Backward propagate via operands
+  if (backward) {
+    LLVM_DEBUG(dbgs() << "backward propagate no contract\n");
+    worklist = roots;
+    while (!worklist.empty()) {
+      auto inst = worklist.pop_back_val();
+      LLVM_DEBUG(dbgs() << "visit: " << *inst << "\n");
+      for (Value *operand : inst->operands()) {
+        if (auto opInst = dyn_cast<Instruction>(operand)) {
+          if (visited.count(opInst))
```

Comment:
Silly little trick: use `!visited.insert(opInst).second` -- this avoids a redundant set lookup (also applies below)

Reply:
Thanks! Implemented.